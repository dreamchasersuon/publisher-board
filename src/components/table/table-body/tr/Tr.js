import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import Td from '../td/Td';
import { ReactComponent as Actions } from '../actions.svg';
import Popover from '../../../popover/Popover';
import { useDispatch } from 'react-redux';
import { setActiveUser } from '../../../../redux/features/settingsFeatureSlice';

export default function Tr({ users, user, handleOnOpenModal }) {
  const dispatch = useDispatch();
  const [activeUserId, setActiveUserId] = useState(null);
  const popoverRef = useRef(null);

  /*
        Open popover using ref,
        when userId was set.
   */
  useEffect(() => {
    if (activeUserId) {
      popoverRef.current.focus();
    }
  }, [activeUserId]);

  /*
        Remove userId, when popover
        is blurred.
   */
  function handleOnBlurPopover() {
    setActiveUserId(null);
  }

  /*
        Set userId, selected by clicking on
        row's popover button.
        Save activeUser to redux to perform
        operations in forms later on.
   */
  function onClickActions(userId) {
    if (userId === activeUserId) {
      return setActiveUserId(null);
    }

    const user = users.find(user => user.user_id === userId);
    dispatch(setActiveUser({ user }));
    setActiveUserId(userId);
  }
  return (
    <tr key={user.user_id}>
      <td>
        <div className="TableBody-cell">
          <div
            className={
              user.enabled ? 'TableBody-enabled' : 'TableBody-disabled'
            }
          />
          {user.enabled ? 'Enabled' : 'Disabled'}
        </div>
      </td>
      <Td value={user.user_name} />
      <Td value={user.user_custom} />
      <Td value={user.email} />
      <Td value={user.balance} />
      <Td isDate value={user.register_date} />
      <td>
        <div
          data-testid="popover-button"
          onClick={() => onClickActions(user.user_id)}
          className="TableBody-cell actions"
        >
          <Actions />
        </div>
        {activeUserId === user.user_id && (
          <Popover
            popoverRef={popoverRef}
            handleOpenModal={handleOnOpenModal}
            handleOnBlurPopover={handleOnBlurPopover}
          />
        )}
      </td>
    </tr>
  );
}

Tr.propTypes = {
  users: PropTypes.array,
  user: PropTypes.object,
  handleOnOpenModal: PropTypes.func,
};
