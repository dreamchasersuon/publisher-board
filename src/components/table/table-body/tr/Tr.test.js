import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Tr from './Tr';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const onOpen = jest.fn();
const user = {
  user_id: 1,
  enabled: true,
  user_name: 'John',
  user_custom: 'Malkovich',
  email: 'saint@malkovich.com',
  balance: '9999999',
  register_date: '2019-01-30T10:24:53+00:00',
};
const users = [
  user,
  {
    user_id: 2,
    enabled: true,
    user_name: 'Charles',
    user_custom: 'Bukovski',
    email: 'buk@charles.com',
    balance: '-999.99',
    register_date: '2019-01-31T10:24:53+00:00',
  },
];

describe('renders table body row', () => {
  test('open popover', async () => {
    const mockStore = configureStore();
    const store = mockStore({});
    const { getByTestId, findByTestId } = render(
      <Provider store={store}>
        <table>
          <tbody>
            <Tr user={user} users={users} handleOnOpenModal={onOpen} />
          </tbody>
        </table>
      </Provider>,
    );

    const btn = getByTestId('popover-button');
    fireEvent.click(btn);

    const popover = await findByTestId('popover');

    expect(popover).toMatchSnapshot();
    expect(store.getActions()).toMatchSnapshot();
  });
});
