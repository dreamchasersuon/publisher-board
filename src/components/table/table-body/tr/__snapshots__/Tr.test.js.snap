// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders table body row open popover 1`] = `
<div
  class="Popover"
  data-testid="popover"
  tabindex="0"
>
  <button
    class="Popover-button"
    data-name="updateUser"
    data-testid="update-user"
  >
    Update User
  </button>
  <button
    class="Popover-button"
    data-name="getTransactions"
    data-testid="get-transactions"
  >
    Get Transactions
  </button>
  <button
    class="Popover-button"
    data-testid="update-balance"
  >
    Update Balance
  </button>
</div>
`;

exports[`renders table body row open popover 2`] = `
Array [
  Object {
    "payload": Object {
      "user": Object {
        "balance": "9999999",
        "email": "saint@malkovich.com",
        "enabled": true,
        "register_date": "2019-01-30T10:24:53+00:00",
        "user_custom": "Malkovich",
        "user_id": 1,
        "user_name": "John",
      },
    },
    "type": "settings/setActiveUser",
  },
]
`;
